File: EmissaoCTe.API/Actions/JsonNetResult.vb
Imports Newtonsoft.Json

Public Class JsonNetResult
    Inherits JsonResult

    Public Property SerializerSettings As JsonSerializerSettings
    Public Property Formatting As Formatting

    Public Sub New()
        SerializerSettings = New JsonSerializerSettings()
        SerializerSettings.Converters.Add(New Newtonsoft.Json.Converters.IsoDateTimeConverter())
    End Sub


    Public Overrides Sub ExecuteResult(context As System.Web.Mvc.ControllerContext)

        If context Is Nothing Then
            Throw New ArgumentNullException("context")
        End If

        Dim response = context.HttpContext.Response

        response.ContentType = If(String.IsNullOrEmpty(ContentType) = False, ContentType, "application/json")

        If ContentEncoding IsNot Nothing Then
            response.ContentEncoding = ContentEncoding
        End If

        If Data IsNot Nothing Then
            Dim writer As New JsonTextWriter(response.Output) With {.Formatting = Formatting}

            Dim serializer = JsonSerializer.Create(SerializerSettings)
            serializer.Serialize(writer, Data)

            writer.Flush()
        End If
    End Sub

End Class

File: EmissaoCTe.API/Actions/JsonNetResult.cs
using System;
using System.Web.Mvc;
using Newtonsoft.Json;

namespace EmissaoCTe.API
{
    public class JsonNetResult : JsonResult
    {

        #region Variaveis Globais

        public JsonSerializerSettings SerializerSettings;
        public Formatting Formatting;

        #endregion

        #region Construtores

        public JsonNetResult()
        {
            this.SerializerSettings = new JsonSerializerSettings();
            this.SerializerSettings.Converters.Add(new Newtonsoft.Json.Converters.IsoDateTimeConverter());
        }

        #endregion

        #region Metodos

        public override void ExecuteResult(ControllerContext context)
        {
            if (context == null)
                throw new ArgumentException("context");

            var response = context.HttpContext.Response;

            response.ContentType = string.IsNullOrWhiteSpace(ContentType) ? "application/json" : ContentType;

            if (ContentEncoding != null)
                response.ContentEncoding = ContentEncoding;

            if (Data != null)
            {
                JsonTextWriter writer = new JsonTextWriter(response.Output) { Formatting = Formatting };

                var serializer = JsonSerializer.Create(SerializerSettings);
                serializer.Serialize(writer, Data);

                writer.Flush();
            }
        }

        #endregion

    }
}